common/src/main/java/org/corfudb/common/util/FileWatcher.java

```diff
@@ -0,0 +1,103 @@
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.time.Duration;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+public class FileWatcher implements Closeable {
+
+    private final File file;
+
+    private final Runnable onChange;
+
+    private final ScheduledExecutorService watcher;
+
+    private volatile WatchService watchService;
+
+
+    public FileWatcher(String filePath, Runnable onChange, Duration pollPeriod){
+        this.file = Paths.get(filePath).toFile();
+        this.onChange = onChange;
+
+        this.watcher = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                        .setNameFormat("FileWatcher-")
+                        .build());
+
+        reloadNewWatchService();
+        watcher.scheduleAtFixedRate(
+                this::poll,
+                0,
+                pollPeriod.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
+
+    private void poll() {
+        try {
+            WatchKey key = watchService.poll();
+            if (key == null) {
+                return;
+            }
+
+            for (WatchEvent<?> event : key.pollEvents()) {
+                WatchEvent.Kind<?> kind = event.kind();
+
+                WatchEvent<Path> ev = (WatchEvent<Path>) event;
+                Path filename = ev.context();
+
+                if (kind == StandardWatchEventKinds.OVERFLOW) {
+                    log.warn("FileWatcher hit overflow and events might be lost!");
+                } else if (kind == StandardWatchEventKinds.ENTRY_MODIFY
+                        && filename.toString().equals(file.getName())) {
+                    log.info("FileWatcher: file {} changed. Invoking handler...", filename);
+                    onChange.run();
+                }
+            }
+            // reset key for continuous watching
+            key.reset();
+        } catch (Throwable t) {
+            log.error("FileWatcher failed to poll file {}", file.getAbsoluteFile(), t);
+            reloadNewWatchService();
+        }
+    }
+
+    private void reloadNewWatchService() {
+        try {
+            if (watchService != null) {
+                watchService.close();
+            }
+            watchService = FileSystems.getDefault().newWatchService();
+            Path path = file.toPath().getParent();
+            path.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY);
+            log.info("FileWatcher: parent dir {} for file {} registered.", path, file.getAbsoluteFile());
+        } catch (IOException ioe) {
+            throw new IllegalStateException("Failed to start a new watch service!", ioe);
```

Comment:
If there's a transient disk failure that triggers an IOException here, then this task will never be rescheduled hereafter.

Reply:
Changed to use `() -> LambdaUtils.runSansThrow(this::poll)` in the scheduled executor.

Reply:
This doesn't seem sufficient. If the IOException occurs on `newWatchService()` then the watcher will remain closed, and in the next iteration of `poll()`, `WatchService::poll` will throw an `ClosedWatchServiceException` and the reload will occur normally. On the other hand, if the IOException is thrown from the call to `Path::register`, the next iteration of `poll()` will not produce such an exception since the `WatchService` will have been reinitialized. In that case, it seems like we won't be able to detect changes to the cert files since the registration never completed successfully.

Reply:
I see your point that the two failure cases can happen and they lead to different failures in the next poll cycle. The hard part is that the potential `IOException` is likely not self-recoverable, and we don't have many solutions other than retrying. What's more, I deem this auto hot-swapping functionality as an add-on to the CorfuRuntime and it seems a bit overreacting if we propagate the error to higher layer and stop the entire CorfuRuntime. In the worst case where the failure is persistent, we can find the error message in log every 60s (not rotated) to identify that the auto-swapping failed due to this IOException.

Reply:
I agree that the failure should not be propagated to the higher layers, but why can't we retry the registration as part of the retry process? It seems like two things need to hold when poll() begins: 1) the WatchService should be initialized, and 2), the appropriate registration should be complete.

In the case where the IOException is persistent, we will be able to find it in the logs as you mentioned. But in the case of a short one-off transient failure, we will have little-to-no way in detecting this scenario if the logs roll over. The watcher thread will simply continue polling but never retrieve any events without any additional logs produced.